File: design_pattern/typestate.ts
// ============================================================
// Typestate Pattern (TypeScript) — 只能运行时检查，对比 Rust 编译期版本
// 对比 Rust: 13_typestate.rs
// 运行: npx ts-node 13_typestate.ts
// ============================================================

type State = "Disconnected" | "Connected" | "Authenticated";

class Client {
  private state: State = "Disconnected";

  constructor(private host: string, private port: number) {}

  connect(): this {
    if (this.state !== "Disconnected") throw new Error(`无法连接：${this.state}`);
    this.state = "Connected";
    console.log(`[${this.host}:${this.port}] 已连接`);
    return this;
  }

  authenticate(token: string): this {
    if (this.state !== "Connected") throw new Error(`无法认证：${this.state}`);
    this.state = "Authenticated";
    console.log(`[${this.host}:${this.port}] 认证成功 token=${token}`);
    return this;
  }

  send(data: string): void {
    if (this.state !== "Authenticated") throw new Error(`无法发送：${this.state}`);
    console.log(`[${this.host}:${this.port}] 发送: ${data}`);
  }

  receive(): string {
    if (this.state !== "Authenticated") throw new Error(`无法接收：${this.state}`);
    const msg = `来自 ${this.host}:${this.port} 的响应`;
    console.log(`[${this.host}:${this.port}] 接收: ${msg}`);
    return msg;
  }
}

// --- main ---
console.log("=== Typestate Pattern (TypeScript) ===\n");

console.log("--- 正确流程 ---");
const client = new Client("api.example.com", 443);
client.connect().authenticate("Bearer xyz");
client.send('{"action":"getUsers"}');
client.receive();

console.log("\n--- 运行时错误（TS 无法在编译期阻止）---");
const bad = new Client("x.com", 80);
try {
  bad.send("data"); // 运行时才抛出！
} catch (e: any) {
  console.log("错误:", e.message);
}

// 两者对比：
// Rust 版本中，未认证的 Client<Disconnected> 根本没有 .send() 方法
// 调用会在编译时报错，程序根本跑不起来
// TS 版本在编译时看起来没问题，只有运行时才发现

File: design_pattern/typestate.rs
// ============================================================
// Typestate Pattern — 用类型系统在编译期强制状态顺序
// 对比 TS: 13_typestate.ts
// 运行: cargo run --bin typestate
// ============================================================

use std::marker::PhantomData;

// 状态标记（零大小类型，纯编译期信息，无运行时开销）
struct Disconnected;
struct Connected;
struct Authenticated;

struct Client<State> {
    host: String,
    port: u16,
    _state: PhantomData<State>,
}

// 只有 Disconnected 状态才有 connect 方法
impl Client<Disconnected> {
    fn new(host: &str, port: u16) -> Self {
        Self { host: host.into(), port, _state: PhantomData }
    }

    fn connect(self) -> Client<Connected> {
        println!("[{}:{}] 已连接", self.host, self.port);
        Client { host: self.host, port: self.port, _state: PhantomData }
    }
}

// 只有 Connected 状态才有 authenticate 方法
impl Client<Connected> {
    fn authenticate(self, token: &str) -> Client<Authenticated> {
        println!("[{}:{}] 认证成功 token={}", self.host, self.port, token);
        Client { host: self.host, port: self.port, _state: PhantomData }
    }

    #[allow(dead_code)]
    fn disconnect(self) -> Client<Disconnected> {
        println!("[{}:{}] 已断开", self.host, self.port);
        Client { host: self.host, port: self.port, _state: PhantomData }
    }
}

// 只有 Authenticated 状态才能 send / receive
impl Client<Authenticated> {
    fn send(&self, data: &str) {
        println!("[{}:{}] 发送: {}", self.host, self.port, data);
    }

    fn receive(&self) -> String {
        let msg = format!("来自 {}:{} 的响应", self.host, self.port);
        println!("[{}:{}] 接收: {}", self.host, self.port, msg);
        msg
    }
}

fn main() {
    println!("=== Typestate Pattern ===\n");

    println!("--- 正确流程 ---");
    let client = Client::<Disconnected>::new("api.example.com", 443);
    let client = client.connect();
    let client = client.authenticate("Bearer xyz");
    client.send(r#"{"action":"getUsers"}"#);
    client.receive();

    println!("\n--- 编译期阻止非法调用（取消注释验证）---");
    // let c = Client::<Disconnected>::new("x.com", 80);
    // c.send("data");
    // ^ error[E0599]: no method named `send` found for `Client<Disconnected>`
    //
    // let c = Client::<Disconnected>::new("x.com", 80).connect();
    // c.send("data");
    // ^ error[E0599]: no method named `send` found for `Client<Connected>`
    //   必须先 authenticate，否则编译不过

    println!("Typestate 保证：未认证的客户端根本没有 send/receive 方法");
    println!("非法调用是编译错误，不是运行时异常");
}

// Rust 关键差异：
// - PhantomData<State> 是零成本的，编译后状态类型不存在于二进制中
// - 状态转换消费 self（move 语义），防止持有多个状态的引用
// - TS 只能运行时 throw，Rust 在编译期就拒绝
